components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/FxaAccountManager.kt

```diff
@@ -0,0 +1,424 @@
+ */
+private interface OAuthObserver {
+    /**
+     * Account manager is requesting for an OAUTH flow to begin.
+     * @param authUrl Starting point for the OAUTH flow.
+     */
+    fun onBeginOAuthFlow(authUrl: String)
+
+    /**
+     * Account manager encountered an error during authentication. Inspect [error] for details.
+     * @param error A specific FxA error encountered.
+     */
+    fun onError(error: FxaException)
+}
+
+/**
+ * Observer interface which lets its users monitor account state changes and major events.
+ */
+interface AccountObserver {
+    /**
+     * Account just got logged out.
+     */
+    fun onLoggedOut()
+
+    /**
+     * Account was successfully authenticated.
+     * @param account An authenticated instance of a [FirefoxAccountShaped].
+     */
+    fun onAuthenticated(account: FirefoxAccountShaped)
+
+    /**
+     * Account's profile is now available.
+     * @param profile A fresh version of account's [Profile].
+     */
+    fun onProfileUpdated(profile: Profile)
+
+    /**
+     * Account manager encountered an error. Inspect [error] for details.
+     * @param error A specific FxA error encountered.
+     */
+    fun onError(error: FxaException)
+}
+
+/**
+ * An account manager which encapsulates various internal details of an account lifecycle and provides
+ * an observer interface along with a public API for interacting with an account.
+ * The internal state machine abstracts over state space as exposed by the fxaclient library, not
+ * the internal states experienced by lower-level representation of a Firefox Account; those are opaque to us.
+ *
+ * Class is 'open' to facilitate testing.
+ *
+ * @param context A [Context] instance that's used for internal messaging and interacting with local storage.
+ * @param config A [Config] used for account initialization.
+ * @param scopes A list of scopes which will be requested during account authentication.
+ */
+open class FxaAccountManager(
+    private val context: Context,
+    private val config: Config,
+    private val scopes: Array<String>,
+    private val accountStorage: AccountStorage = SharedPrefAccountStorage(context)
+) : Closeable, Observable<AccountObserver> by ObserverRegistry() {
+    private val logTag = "FirefoxAccountStateMachine"
+
+    private val oauth = object : Observable<OAuthObserver> by ObserverRegistry() {}
+
+    companion object {
+        /**
+         * State transition matrix. It's in the companion object to enforce purity.
+         * @return An optional [AccountState] if provided state+event combination results in a
+         * state transition. Note that states may transition into themselves.
+         */
+        @Suppress("ComplexMethod")
+        internal fun nextState(state: AccountState, event: SMEvent): AccountState? {
+            return when (state) {
+                AccountState.Start -> {
+                    when (event) {
+                        SMEvent.Init -> AccountState.Start
```

Comment:
nit (don't change if nothing else needs an update): since this is private I would probably just go with `Event`. Not sure it's obvious what SM stands for :)